README.md
<a href="/README_es.md">Readme en Español</a>

# Password generator

## Introduction

This is a simple app that generates random strong passwords.

### Starting

You can download the executable of the version you want.

### Pre-requisites 📋

-Have the minimum necessary Java jre installed, the [jre1.8.0_321](https://www.java.com/es/download/ie_manual.jsp).

### Installation 🔧

Download <a href="https://github.com/JSalmon11/Generador-de-Contrasenias/releases">here</a> the last version.

With Java on your computer you will only have to run the .exe or .jar file.

## Execution ⚙️

When executing the program, the desired length of the password is requested, once chosen, press the show button and a secure password will appear, press copy to put it on the clipboard and be able to paste it wherever you want.

### Author ✍️
https://github.com/JSalmon11

## Built with 🛠️

* [Visual studio code](https://code.visualstudio.com/download)


CHANGELOG.md
# CHANGELOG

## [1.2.0] - 2023-06-24
### Added
- Added verification not to repeat characters more than 2 times when generating the password.
- Added check not to repeat consecutive identical characters.
- Added verification to not repeat consecutive uppercase characters.
- Added check not to repeat consecutive lowercase characters.
- Added verification not to repeat consecutive numeric characters.

## [1.1.1] - 2022-03-12
### Fixed
- Fixed button to check available updates.

## [1.1.0] - 2022-04-10
### Added
- Added button to check available updates.
- Added internationalization support  (es, en, ja).

### Fixed
- Handled error when length field is empty.


## [1.0.0] - 2022-04-08
- Initial release

[1.2.0]: https://github.com/JSalmon11/Generador-de-Contrasenias/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/JSalmon11/Generador-de-Contrasenias/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/JSalmon11/Generador-de-Contrasenias/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/JSalmon11/Generador-de-Contrasenias/compare/b552912c04419d8d2ae5d9fd9bb7f315145f46d8...1.0.0

src/i18n/Idioma.java
package i18n;

import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.Properties;

/**
 * @author <a href=
 *         "https://www.discoduroderoer.es/como-hacer-una-aplicacion-multi-idioma-en-java/">Disco
 *         duro de roer</a>
 */
public class Idioma extends Properties {

    private static final long serialVersionUID = 1L;

    /**
     * Carga el idioma deseado y traduce la aplicación.
     * 
     * @author <a href=
     *         "https://www.discoduroderoer.es/como-hacer-una-aplicacion-multi-idioma-en-java/">Disco
     *         duro de roer</a>
     * @param idioma String que contiene el idioma del usuario.
     */
    public Idioma(String idioma) {
        switch (idioma) {
            case "es":
                getProperties("espanol.properties");
                break;
            case "en":
                getProperties("ingles.properties");
                break;
            case "ja":
                getProperties("japanese.properties");
                break;
            default:
                getProperties("ingles.properties");
        }

    }

    private void getProperties(String idioma) {
        try {
            this.load(new InputStreamReader(getClass().getResourceAsStream(idioma), Charset.forName("UTF-8")));
        } catch (IOException ex) {

        }
    }
}

src/Utils/Windows.java
package Utils;

import java.io.IOException;
import java.io.InputStream;
import java.io.StringWriter;

/**
 * Clase para diferentes utilidades en Windows.
 */
public class Windows {
    private static final String REGQUERY_UTIL = "reg query ";
    private static final String REGDWORD_TOKEN = "REG_DWORD";
    private static final String DARK_THEME_CMD = REGQUERY_UTIL
            + "\"HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize\""
            + " /v AppsUseLightTheme";

    /**
     * Método para comprobar el tema activo en Windows.
     * @author <a href="https://gist.github.com/HanSolo/7cf10b86efff8ca2845bf5ec2dd0fe1d">Gerrit Grunwald</a>
     * @return
     *         <ul>
     *         <li>True si está activado el modo oscuro de Windows.</li>
     *         <li>False si está desactivado el modo oscuro de Windows.</li>
     *         </ul>
     */
    public static boolean isWindowsDarkMode() {
        try {
            Process process = Runtime.getRuntime().exec(DARK_THEME_CMD);
            StreamReader reader = new StreamReader(process.getInputStream());

            reader.start();
            process.waitFor();
            reader.join();

            String result = reader.getResult();
            int p = result.indexOf(REGDWORD_TOKEN);

            if (p == -1) {
                return false;
            }

            // 1 == Light Mode, 0 == Dark Mode
            String temp = result.substring(p + REGDWORD_TOKEN.length()).trim();
            return ((Integer.parseInt(temp.substring("0x".length()), 16))) == 0;
        } catch (Exception e) {
            return false;
        }
    }

    static class StreamReader extends Thread {
        private InputStream is;
        private StringWriter sw;

        StreamReader(InputStream is) {
            this.is = is;
            sw = new StringWriter();
        }

        public void run() {
            try {
                int c;
                while ((c = is.read()) != -1)
                    sw.write(c);
            } catch (IOException e) {
            }
        }

        String getResult() {
            return sw.toString();
        }
    }

}


src/Utils/GeneratePass.java
package Utils;

import java.util.concurrent.ThreadLocalRandom;

/**
 * Genera una contraseña segura.
 * 
 * @author <a href="https://github.com/JSalmon11">Jorge Salmón</a>
 */
public class GeneratePass {
    /**
     * Genera una contraseña con las especificaciones deseadas por el usuario.
     * 
     * @author <a href="https://github.com/JSalmon11">Jorge Salmón</a>
     * @param longitud longitud de la contraseña.
     * @return String de una contraseña segura.
     */
    public static String generarContraseñaSegura(int longitud) {
        if (longitud == -1) {
            return "";
        } else {
            StringBuilder sb = new StringBuilder(longitud);
            char c;
            char lastChar = '\u0000'; //lastChar valor null
            for (int i = 0; i < sb.capacity(); ++i) {
                boolean repeat;
                int count;
                do {
                    repeat = false;
                    count = 0;
                    c = (char) ThreadLocalRandom.current().nextInt(33, 126 + 1);
                    for (int j = 0; j < i; j++) {
                        if (sb.charAt(j) == c) {
                            count++;
                        }
                        if (count >= 2) {
                            repeat = true;
                            break;
                        }
                    }
                    if (c == lastChar) {
                        repeat = true;
                    } else if (Character.isLetter(c) && Character.isLetter(lastChar)) {
                        if (Character.isUpperCase(c) == Character.isUpperCase(lastChar)) {
                            repeat = true;
                        }
                    } else if (Character.isDigit(c) && Character.isDigit(lastChar)) {
                        repeat = true;
                    }
                } while (repeat);
                sb.insert(i, c);
                lastChar = c;
            }
            return sb.toString();
        }
    }

}

src/Utils/FormActualizar.java
package Utils;

import java.io.IOException;
import java.awt.Desktop;
import java.net.URI;
import java.net.URISyntaxException;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.layout.Pane;
import javafx.scene.layout.StackPane;
import javafx.scene.text.Font;
import javafx.scene.text.Text;
import javafx.stage.Stage;
import javafx.stage.StageStyle;
import Main.App;
import i18n.Idioma;

public class FormActualizar {
    private static final Idioma idioma = App.idioma;

    /**
     * Abre un nuevo formulario para elegir si actualizar o no a la nueva verión
     * disponible.
     * 
     * @author <a href="https://github.com/JSalmon11">Jorge Salmón</a>
     * @param nuevaVersion String que lleva la URL de la nueva verión disponible.
     */
    public static void actualizar(String nuevaVersion, Stage stage) {
        if (nuevaVersion.equals("-1")) {
            notUpdate(stage, idioma.getProperty("textActualizar"));
        } else if (nuevaVersion.equals("-2")){
            notUpdate(stage, ComprobarActualizaciones.error);
        } else {
            stage.setTitle(idioma.getProperty("actualizarTitulo"));
            stage.initStyle(StageStyle.UTILITY);

            Button actualizar = new Button(idioma.getProperty("buttonActualizar"));
            actualizar.setLayoutX(20);
            actualizar.setLayoutY(20);

            Button cancelar = new Button(idioma.getProperty("buttonCancelar"));
            cancelar.setLayoutX(215);
            cancelar.setLayoutY(20);
            Pane datos = new Pane();
            datos.getChildren().addAll(actualizar, cancelar);
            Scene scene = new Scene(new StackPane(datos), 300, 50);
            if (Windows.isWindowsDarkMode()) {
                scene.getStylesheets().add("visual/dark-mode.css");
            }

            stage.setScene(scene);
            stage.show();

            actualizar.setOnAction(new EventHandler<ActionEvent>() {
                @Override
                public void handle(ActionEvent e) {
                    if (Desktop.isDesktopSupported()) {
                        Desktop desktop = Desktop.getDesktop();
                        try {
                            desktop.browse(new URI(nuevaVersion));
                        } catch (IOException | URISyntaxException ex) {
                            ex.printStackTrace();
                        }
                    } else {
                        Runtime runtime = Runtime.getRuntime();
                        try {
                            runtime.exec("xdg-open " + nuevaVersion);
                        } catch (IOException ex) {
                            ex.printStackTrace();
                        }
                    }
                }
            });

            cancelar.setOnAction(new EventHandler<ActionEvent>() {
                @Override
                public void handle(ActionEvent e) {
                    stage.close();
                }
            });
        }
    }

    private static void notUpdate(Stage stage, String error) {
        stage.setTitle(idioma.getProperty("actualizarTitulo"));
        stage.initStyle(StageStyle.UTILITY);

        Text textActualizar = new Text(error);
        textActualizar.setFont(new Font("Roboto", 14));
        textActualizar.setLayoutX(14);
        textActualizar.setLayoutY(5);

        Scene scene = new Scene(new StackPane(textActualizar), 300, 50);
        if (Windows.isWindowsDarkMode()) {
            scene.getStylesheets().add("visual/dark-mode.css");
        }
        
        stage.setScene(scene);
        stage.show();
    }

}

src/Utils/ComprobarActualizaciones.java
package Utils;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Scanner;

import Main.App;
import i18n.Idioma;

public class ComprobarActualizaciones {
    private static final Idioma idioma = App.idioma;
    static String error = null;

    public String getErrorMessage() {
        return error;
    }

    public static void setErrorMessage(String msg) {
        error = msg;
    }

    /**
     * Comprueba si hay una nueva versión disponible de la aplicación mediante
     * GitHub.
     * 
     * @author <a href="https://github.com/JSalmon11">Jorge Salmón</a>
     * @param version Versión actual de la aplicación.
     * @return
     *         <ul>
     *         <li>String de la nueva versión de la aplicación.</li>
     *         <li>String "-1" si no hay nueva versión disponible.</li>
     *         <li>String "-2" si ha surgido algún error.</li>
     *         </ul>
     */
    public static String checkUpdate(String version) {
        String tagVersion = getLastTag();

        if (!tagVersion.equals("-2")) {
            if (Integer.parseInt(version.replace(".", "")) >= Integer.parseInt(tagVersion.replace(".", ""))) {
                tagVersion = "-1";
            }
        }

        return tagVersion;
    }

    public static String getLastTag() {
        try {
            URL url = new URL("https://api.github.com/repos/JSalmon11/Generador-de-Contrasenias/tags");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.connect();
            int responseCode = conn.getResponseCode();

            if (responseCode != 200) {
                setErrorMessage(responseCode + idioma.getProperty("textErrorGitHub"));
                return "-2";
            } else {
                StringBuilder inline = new StringBuilder();
                Scanner scanner = new Scanner(url.openStream());
                while (scanner.hasNext()) {
                    inline.append(scanner.nextLine());
                }
                scanner.close();

                return inline.substring(10, 15);
            }
        } catch (IOException e) {
            setErrorMessage(idioma.getProperty("textErrorInteno"));
        }
        return "-2";
    }

}

src/Main/App.java
package Main;

import javafx.application.Application;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.TextField;
import javafx.scene.control.Tooltip;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import javafx.scene.layout.StackPane;
import javafx.scene.text.Font;
import javafx.scene.text.Text;
import javafx.stage.Stage;
import java.awt.datatransfer.StringSelection;
import Utils.ComprobarActualizaciones;
import Utils.FormActualizar;
import Utils.GeneratePass;
import Utils.Windows;
import i18n.Idioma;

import java.awt.Toolkit;
import java.awt.datatransfer.Clipboard;

/**
 * Crea y ejecuta una interfaz de usuario para generar contraseñas seguras.
 * 
 * @author <a href="https://github.com/JSalmon11">Jorge Salmón</a>
 */
public class App extends Application {
    private static String version = "1.2.0";
    public static final Idioma idioma = new Idioma(System.getProperty("user.language"));

    public static void main(String[] args) {
        launch();
    }

    @Override
    public void start(Stage stage) {
        stage.getIcons().add(new Image(App.class.getResourceAsStream("/img/icon.png")));
        stage.setResizable(false);
        stage.setTitle(idioma.getProperty("mainTitulo"));

        Text textLongitud = new Text(idioma.getProperty("textLongitud"));
        textLongitud.setFont(new Font("Roboto", 20));
        textLongitud.setLayoutX(14);
        textLongitud.setLayoutY(38);

        TextField longitud = new TextField("14");
        longitud.setMaxSize(50, 14);
        longitud.setFont(new Font("Roboto", 20));
        longitud.setLayoutX(110);
        longitud.setLayoutY(10);

        TextField contraseñas = new TextField();
        contraseñas.setMinSize(272, 20);
        contraseñas.setLayoutX(14);
        contraseñas.setLayoutY(290);

        Button mostrar = new Button(idioma.getProperty("buttonMostrar"));
        mostrar.setLayoutX(14);
        mostrar.setLayoutY(340);

        Button copiar = new Button(idioma.getProperty("buttonCopiar"));
        copiar.setLayoutX(235);
        copiar.setLayoutY(340);

        Button actualizar = new Button();
        actualizar.setLayoutX(270);
        actualizar.setLayoutY(5);
        actualizar.setPrefSize(10, 10);
        actualizar.setGraphic(new ImageView(new Image(App.class.getResourceAsStream("/img/buttons/update.png"))));
        actualizar.setTooltip(new Tooltip(idioma.getProperty("hoverButtonActualizar")));

        Pane datos = new Pane();
        datos.getChildren().addAll(longitud, mostrar, copiar, actualizar, textLongitud, contraseñas);

        Scene scene = new Scene(new StackPane(datos), 300, 380);
        if (Windows.isWindowsDarkMode()) {
            scene.getStylesheets().add("visual/dark-mode.css");
        }

        stage.setScene(scene);
        stage.show();

        mostrar.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent e) {
                int intLongitud = -1;
                try {
                    intLongitud = Integer.parseInt(longitud.getText());
                } catch (NumberFormatException ex) {
                }
                contraseñas.setText((GeneratePass.generarContraseñaSegura(intLongitud)));
            }
        });

        copiar.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent e) {
                Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
                clipboard.setContents(new StringSelection(contraseñas.getText()), null);
            }
        });

        actualizar.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent e) {
                String versionUrl = ComprobarActualizaciones.checkUpdate(version);
                if (!versionUrl.equals("-1") && !versionUrl.equals("-2")) {
                    String nuevaVersion;
                    nuevaVersion = "https://github.com/JSalmon11/Generador-de-Contrasenias/releases/tag/" + versionUrl;
                    Stage updateStage = new Stage();
                    FormActualizar.actualizar(nuevaVersion, updateStage);
                } else {
                    Stage updateStage = new Stage();
                    FormActualizar.actualizar(versionUrl, updateStage);
                }
            }
        });

    }

}